docs/tradeoffs.md

```diff
@@ -0,0 +1,136 @@
+  definition order)
+
+While `bpaf` supports fewer use case / CLI styles, it has an overall simpler
+parser that then scales up in runtime, build time, and code size based on the
+arguments themselves, only paying for what you use.  In addition, some
+context-sensitive cases can be reproduced using the argument combinators (e.g.
+`--color [WHEN]`).
+
+## Validation
+
+Specifically when arguments conflict, override, or require each other.
+
+`bpaf` uses function/macro combinators to declare argument / group of argument relationships.  This
+offers a lot of flexibility that, again, you only pay for what you use.
+
+The downside to this approach is it couples together:
+- Parsing disambiguation with positionals
+- Validation (in a strict tree structure)
+- Help ordering
+- Help sections headers
+- Code structure building up the combinators
+- Data structures as each level is its own `struct`
+- Organizing some of the validation rules aren't always as intuitive for people
+  not steeped in a functional way of thinking
+
+`clap` provides `ArgGroup` to compose arguments and other groups with
+relationships defined in terms of either group or argument IDs.  `ArgGroup`s
+(and the separate help section header feature) are tags on arguments.  This
+allows a very flexible directed acyclic graph of relationships.
+
+The downside to this approach
+- Everyone pays the runtime, compile time, and code size cost, no matter which subset they are using
+- Developers are limited by what relationships `clap` has predefined
+- Even once `clap` opens up to user-provided relationships, the ergonomics for
+  defining them won't be as nice as it will require implementing a trait that
+  uses the lower levels of clap's API and then passing it in to the parser
+
+## Derive APIs
+
+Both libraries provide derive APIs that mask over the static and dynamic typing differences.
+
+In `bpaf`s case, the combinators still show through in terms of requiring the
+user to organize their data structures around their validation.  Some times
+this is good (pushing errors to compile time like if mutually exclusive
+arguments are represented in an `enum`) while at other times it has the potential to convolute the code.
+
+In `clap`s case, it has the challenge of hand-implemented support to express
+each case of argument relationships in the type system (which hasn't been done
+yet).
+
+In both cases, some errors are still pushed off from compile time to early test
+time through asserts.
+
+## Maturity
+
+While this document is focused on design trade-offs, we understand some users
+will look to this to help understand which would work better for them.
+
+`clap` has been around for many more years and has a lot more users from
+varying backgrounds solving different problems and `clap` has taken input and
+adapted to help meet a variety of needs.
+
+`bpaf` is a younger project but it is able to move a lot more quickly because
```

Comment:
> I'm not sure what the ticket is about (probably needs more careful reading)

Feel free to ping me

> ticket asks for non opengroup-like thing (bash +O shopt)

For clap, we are looking at solving this in two ways
- We've split out the lexer and are looking to make it pluggable
- We are opening up the argument ids (for us they are strings) so that a custom lexer could include the prefix and the user would then include that prefix in the argument id to get a match

> Maybe I should go though closed tickets for more interesting examples, again - once I get my computer back and polish/publish current WIP stuff.

I'd recommend also branching out to see what other situations are.

Also, give a try emulating some more complex use cases, like more of git (diff's weird use of `--` for "always last argument, stash's default subcommands, etc) or cargo with `cargo run` (can't think of other cases).

Reply:
For non-opengroup things - I think it's going to be a user facing preprocessor - it should also cover some more strange usercases but keep the help generation in terms of processed names, so `+O` is accepted but help would be talking about `-O` unless there's some custom overrides for that which sould be already possible.

Will take a look at `git`, thanks for the tip. As for `cargo run` - I think I already implemented it as a part of unreleased https://github.com/pacak/bpaf/tree/master/bpaf_cauwugo - this gets dynamic completion available for workspace binaries/examples without having to fight `cargo` and mess with manually installing completions for all the things. Plus some other convenience things. But that's still some time away.